src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java

```diff
@@ -117,39 +119,114 @@
-            ? newProductVariantImages.stream()
-                                     .collect(Collectors.toMap(Image::getUrl, image -> image)) : Collections.emptyMap();
+        // this implementation is quite straight forward and might be slow on large arrays,
+        // because of quadratic algorithms of remove/add images synchronization.
+        // Unfortunately, there is not easy solution to synchronize 2 ordered lists
+        // having only add/remove/moveToPos actions.
+        // This solution should be re-optimized in the next releases to avoid O(N^2) for large lists.
+
+        if (!Objects.equals(oldProductVariantImages, newProductVariantImages)) {
+            final List<Image> oldImages = oldProductVariantImages != null
+                    ? oldProductVariantImages : Collections.emptyList();
+
+            final List<Image> updatedOldImages = new ArrayList<>(oldImages);
+
+            final List<Image> newImages = newProductVariantImages != null
+                    ? newProductVariantImages : Collections.emptyList();
+
+            filterCollection(oldProductVariantImages, oldVariantImage ->
+                    !newImages.contains(oldVariantImage))
+                    .forEach(oldImage -> {
+                        updateActions.add(RemoveImage.ofVariantId(oldProductVariantId, oldImage, true));
+                        updatedOldImages.remove(oldImage);
+                    });
+
+            filterCollection(newProductVariantImages, newVariantImage ->
+                    !oldImages.contains(newVariantImage))
+                    .forEach(newImage -> {
+                        updateActions.add(AddExternalImage.ofVariantId(oldProductVariantId, newImage, true));
+                        updatedOldImages.add(newImage);
+                    });
+            updateActions.addAll(buildMoveImageToPositionUpdateActions(oldProductVariantId,
+                    updatedOldImages, newImages));
+        }
+        return updateActions;
+    }
 
-        final Integer oldProductVariantId = oldProductVariant.getId();
+    /**
+     * Compares an old {@link List} of {@link Image}s and a new one and returns a {@link List} of
+     * {@link MoveImageToPosition}&lt;{@link Product}&gt; with the given {@code variantId}. If both the lists are
+     * identical, then no update action is needed and hence an empty {@link List} is returned.
+     *
+     * <p>This method expects the two lists two contain the same images only in different order. Therefore, be cautios
+     * that supplying lists of different (missing/extra) images could results in an index out of bounds exception on the
+     * new position of an image.
+     *
+     * <p><b>Note</b>: the solution is still not optimized and may contain {@link MoveImageToPosition} actions
+     * for items which are already on desired positions (after previous moves in the sequence). This will be
+     * re-optimized in the next releases.
+     *
```

Comment:
i put my comments here, pls read them and make your notes (if necessary)

Reply:
I reworded the JavaDoc a bit and added a GH ISSUE TODO: https://github.com/commercetools/commercetools-sync-java/issues/133